src/systems/physics/Physics.cc

```diff
@@ -1726,51 +1734,62 @@ void PhysicsPrivate::UpdateSim(EntityComponentManager &_ecm)
         auto frameData = linkPhys->FrameDataRelativeToWorld();
         const auto &worldPose = frameData.pose;
 
-        if (canonicalLink)
-        {
-          // This is the canonical link, update the top level model.
-          // The pose of this link w.r.t its top level model never changes
-          // because it's "fixed" to the model. Instead, we change
-          // the top level model's pose here. The physics engine gives us the
-          // pose of this link relative to world so to set the top level
-          // model's pose, we have to post-multiply it by the inverse of the
-          // transform of the link w.r.t to its top level model.
-          math::Pose3d linkPoseFromTopLevelModel;
-          linkPoseFromTopLevelModel =
-              this->RelativePose(topLevelModelEnt, _entity, _ecm);
-
-          // update top level model's pose
-          auto mutableModelPose =
-             _ecm.Component<components::Pose>(topLevelModelEnt);
-          *(mutableModelPose) = components::Pose(
-              math::eigen3::convert(worldPose) *
-              linkPoseFromTopLevelModel.Inverse());
-
-          _ecm.SetChanged(topLevelModelEnt, components::Pose::typeId,
-              ComponentState::PeriodicChange);
-        }
-        else
+        // update the link or top level model pose if this is the first update,
+        // or if the link pose has changed since the last update
+        // (if the link pose hasn't changed, there's no need for a pose update)
+        if ((this->linkWorldPoses.find(_entity) == this->linkWorldPoses.end())
+            || (this->linkWorldPoses[_entity].matrix() != worldPose.matrix()))
```

Comment:
I took a look at the documentation for [`Eigen::Transform`](https://eigen.tuxfamily.org/dox/classEigen_1_1Transform.html) (`ignition::Physics::Pose` is really `Eigen::Transform` [under the hood](https://github.com/ignitionrobotics/ign-physics/blob/6aa968042659ad4ee60e9126a7ff81fe68d0d05c/include/ignition/physics/Geometry.hh#L44)), and it doesn't seem to have an `==` operator defined. So, in order to see if the transforms are equal, I had to call the transform's [`matrix()` method](https://eigen.tuxfamily.org/dox/classEigen_1_1Transform.html#a758fbaf6aa41a5493659aa0c9bfe0dcf) and then compare the matrices for equality. Is this okay, or is there a better way of checking if the two transforms are equal?

Reply:
Comparing the matrices directly is likely doing comparison on doubles, which means the matrices have to be exactly the same for equality. This might be okay for the purpose of this code because it only causes an update if the matrices aren't equal. A different way would be to compare the norm of the difference of the translations with some tolerance and then convert the relative rotation to angle axis representation and check the angle (https://github.com/ignitionrobotics/ign-physics/blob/9ba272864ec261ffe972365c82a808b476628659/test/Utils.hh#L181)

Since we have comparison functions for `ignition::math::Pose3d` in Physics.cc already, it might be easier to store them as `ignition::math::Pose3d`. You can do the conversion with `ignition::math::eigen3::convert`.

Reply:
> Comparing the matrices directly is likely doing comparison on doubles, which means the matrices have to be exactly the same for equality.

You're right; I looked into this some more, and the [MatrixBase == operator](https://eigen.tuxfamily.org/dox/classEigen_1_1MatrixBase.html#a80e3e1e83fdf43f9f7fb6ff51836b24d) has no tolerance for floats. While we could use [`isApprox()`](https://eigen.tuxfamily.org/dox/classEigen_1_1DenseBase.html#ae8443357b808cd393be1b51974213f9c), it looks like there could be some issues with zero vectors/matrices.

> Since we have comparison functions for `ignition::math::Pose3d` in `Physics.cc` already, it might be easier to store them as `ignition::math::Pose3d`. You can do the conversion with `ignition::math::eigen3::convert`.

I went ahead and used `ignition::math::Pose3d` and the corresponding comparison functions in `Physics.cc` in https://github.com/ignitionrobotics/ign-gazebo/pull/669/commits/6bf05727f38212c16f9b4ac3e26b6d5058f005d8. Another nice thing about using `ignition::math::Pose3d` is that it removes the hardcoded values from `physics::Pose` that [I mentioned above](https://github.com/ignitionrobotics/ign-gazebo/pull/669#discussion_r587785042). If this looks good to you, I think we can mark this as resolved.